File: src/scripts/fetch-gov-news.ts
import { initializeApp } from 'firebase/app'
import { getAuth, signInWithEmailAndPassword } from 'firebase/auth'
import { getFirestore, collection, addDoc, query, where, getDocs, Timestamp, limit } from 'firebase/firestore'
import dotenv from 'dotenv'
import { resolve } from 'path'
import { load } from 'cheerio'
import * as readline from 'readline'

// 載入環境變量
dotenv.config({ path: resolve(process.cwd(), '.env') })

// 初始化 Firebase
const firebaseConfig = {
  apiKey: process.env.VITE_FIREBASE_API_KEY,
  authDomain: process.env.VITE_FIREBASE_AUTH_DOMAIN,
  projectId: process.env.VITE_FIREBASE_PROJECT_ID,
  storageBucket: process.env.VITE_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.VITE_FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.VITE_FIREBASE_APP_ID,
}

if (!firebaseConfig.projectId) {
  throw new Error('請確保 .env 文件中已設置所有 Firebase 配置變量')
}

console.log(`📋 使用 Firebase 項目: ${firebaseConfig.projectId}`)

const app = initializeApp(firebaseConfig)
const auth = getAuth(app)
const db = getFirestore(app)

// 創建 readline 接口
function createReadlineInterface() {
  return readline.createInterface({
    input: process.stdin,
    output: process.stdout
  })
}

// 從命令行參數或環境變量獲取認證信息
function getAuthFromArgs(): { email: string; password: string } | null {
  const email = process.env.ADMIN_EMAIL || process.argv[2]
  const password = process.env.ADMIN_PASSWORD || process.argv[3]
  
  if (email && password) {
    return { email, password }
  }
  return null
}

// 提示用戶輸入認證信息
function promptForAuth(): Promise<{ email: string; password: string }> {
  return new Promise((resolve) => {
    const rl = createReadlineInterface()
    
    rl.question('請輸入管理員電子郵件: ', (email) => {
      rl.question('請輸入管理員密碼: ', (password) => {
        rl.close()
        resolve({ email, password })
      })
    })
  })
}

// 認證用戶
async function authenticate(): Promise<void> {
  console.log('\n🔐 需要管理員認證才能添加公告...\n')
  
  let credentials = getAuthFromArgs()
  
  if (!credentials) {
    credentials = await promptForAuth()
  }
  
  try {
    console.log('正在登入...')
    await signInWithEmailAndPassword(auth, credentials.email, credentials.password)
    console.log('✅ 認證成功\n')
  } catch (error: any) {
    console.error('\n❌ 認證失敗:', error.message)
    throw error
  }
}

// 火災相關關鍵詞（核心關鍵詞，必須包含）
const CORE_FIRE_KEYWORDS = [
  '火',
  '火警',
  '火災',
  '火災事故',
  '火災現場',
  '宏福苑', // 特定地點
]

// 輔助關鍵詞（如果與核心關鍵詞一起出現，則更可能是相關的）
const SUPPORTING_KEYWORDS = [
  '大埔',
  '宏福',
  '庇護中心',
  '臨時庇護',
  '疏散',
  '消防',
  '救援',
  '緊急',
  '撤離',
]

// 檢查文本是否與火災相關
function isFireRelated(text: string): boolean {
  if (!text || text.trim().length === 0) {
    return false
  }
  
  const lowerText = text.toLowerCase()
  
  // 必須包含至少一個核心關鍵詞
  const hasCoreKeyword = CORE_FIRE_KEYWORDS.some(keyword =>
    lowerText.includes(keyword.toLowerCase())
  )
  
  if (hasCoreKeyword) {
    return true
  }
  
  // 如果沒有核心關鍵詞，檢查是否同時包含多個輔助關鍵詞
  // 這可以幫助過濾掉只包含"緊急"或"救援"但與火災無關的新聞
  const supportingCount = SUPPORTING_KEYWORDS.filter(keyword =>
    lowerText.includes(keyword.toLowerCase())
  ).length
  
  // 如果包含 2 個或以上的輔助關鍵詞，且包含"大埔"或"宏福"，則認為相關
  if (supportingCount >= 2) {
    return lowerText.includes('大埔') || lowerText.includes('宏福')
  }
  
  return false
}

// 解析 RSS pubDate 為中文日期格式
function parseRSSDate(pubDate: string): string {
  try {
    const date = new Date(pubDate)
    const year = date.getFullYear()
    const month = date.getMonth() + 1
    const day = date.getDate()
    return `${year}年${month}月${day}日`
  } catch (error) {
    return new Date().toLocaleDateString('zh-HK')
  }
}

// 清理 HTML 標籤和實體
function cleanHtml(html: string): string {
  return html
    .replace(/<[^>]*>/g, '') // 移除 HTML 標籤
    .replace(/&nbsp;/g, ' ') // 替換 &nbsp;
    .replace(/&amp;/g, '&') // 替換 &amp;
    .replace(/&lt;/g, '<') // 替換 &lt;
    .replace(/&gt;/g, '>') // 替換 &gt;
    .replace(/&quot;/g, '"') // 替換 &quot;
    .replace(/&#39;/g, "'") // 替換 &#39;
    .replace(/\s+/g, ' ') // 合併多個空格
    .trim()
}

// 獲取政府新聞公報（使用 RSS Feed）
async function fetchGovNews(): Promise<Array<{ title: string; url: string; date: string; description?: string; pubDate?: string }>> {
  const rssUrl = 'https://www.info.gov.hk/gia/rss/general_zh.xml'
  
  try {
    console.log(`📰 正在從 RSS Feed 獲取政府新聞: ${rssUrl}`)
    
    const response = await fetch(rssUrl, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
      }
    })
    
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }
    
    const xml = await response.text()
    const $ = load(xml, { xmlMode: true })
    
    const newsItems: Array<{ title: string; url: string; date: string; description?: string; pubDate?: string }> = []
    
    // 解析 RSS items
    $('item').each((_, element) => {
      const $item = $(element)
      const title = $item.find('title').text().trim()
      const link = $item.find('link').text().trim()
      const pubDate = $item.find('pubDate').text().trim()
      const description = $item.find('description').text().trim()
      
      if (!title || !link) {
        return
      }
      
      // 檢查是否與火災相關（檢查標題和描述）
      const titleRelated = isFireRelated(title)
      const descRelated = isFireRelated(description)
      
      if (titleRelated || descRelated) {
        const date = parseRSSDate(pubDate)
        console.log(`✅ 找到相關新聞: ${title}`)
        newsItems.push({
          title,
          url: link,
          date,
          description: cleanHtml(description),
          pubDate: pubDate || undefined // 保留原始 pubDate 用於時間戳解析
        })
      } else {
        console.log(`⏭️  跳過不相關新聞: ${title}`)
      }
    })
    
    console.log(`✅ 從 RSS Feed 找到 ${newsItems.length} 條相關新聞\n`)
    return newsItems
  } catch (error: any) {
    console.error(`❌ 獲取 RSS Feed 時發生錯誤: ${error.message}`)
    throw new Error(`無法獲取政府新聞 RSS Feed: ${error.message}`)
  }
}

// 獲取新聞詳細內容
async function fetchNewsContent(url: string): Promise<string> {
  try {
    const response = await fetch(url)
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }
    
    const html = await response.text()
    const $ = load(html)
    
    // 提取新聞內容（根據實際 HTML 結構調整選擇器）
    let content = ''
    
    // 嘗試多種可能的內容選擇器
    const contentSelectors = [
      '#pressrelease',
      '.pressrelease',
      '#content',
      '.content',
      'article',
      'main'
    ]
    
    for (const selector of contentSelectors) {
      const $content = $(selector)
      if ($content.length > 0) {
        content = $content.text().trim()
        break
      }
    }
    
    // 如果找不到特定容器，嘗試獲取所有段落
    if (!content) {
      $('p').each((_, element) => {
        const text = $(element).text().trim()
        if (text.length > 20) { // 過濾太短的段落
          content += text + '\n\n'
        }
      })
    }
    
    return content.trim() || '無法獲取新聞內容'
  } catch (error: any) {
    console.error(`❌ 獲取新聞內容時發生錯誤 (${url}):`, error.message)
    return '無法獲取新聞內容'
  }
}

// 檢查新聞是否已存在
async function newsExists(title: string, url: string): Promise<boolean> {
  try {
    // 檢查標題或 URL 是否已存在
    const titleQuery = query(
      collection(db, 'news'),
      where('title', '==', title),
      limit(1)
    )
    const titleSnapshot = await getDocs(titleQuery)
    if (!titleSnapshot.empty) {
      return true
    }

    const urlQuery = query(
      collection(db, 'news'),
      where('url', '==', url),
      limit(1)
    )
    const urlSnapshot = await getDocs(urlQuery)
    return !urlSnapshot.empty
  } catch (error) {
    console.error('檢查新聞是否存在時發生錯誤:', error)
    return false
  }
}

// 簡單的備用分類（關鍵詞匹配）
function classifyNewsFallback(title: string, content: string): string {
  const text = `${title} ${content}`.toLowerCase()
  
  if (text.includes('火勢') || text.includes('救援') || text.includes('現場') || text.includes('進展')) {
    return 'event-update'
  }
  if (text.includes('資助') || text.includes('補助') || text.includes('津貼') || text.includes('賠償')) {
    return 'financial-support'
  }
  if (text.includes('心理') || text.includes('輔導') || text.includes('情緒')) {
    return 'emotional-support'
  }
  if (text.includes('庇護') || text.includes('住宿') || text.includes('臨時')) {
    return 'accommodation'
  }
  if (text.includes('醫療') || text.includes('法律')) {
    return 'medical-legal'
  }
  if (text.includes('重建')) {
    return 'reconstruction'
  }
  if (text.includes('死亡') || text.includes('受傷') || text.includes('失蹤') || text.includes('統計')) {
    return 'statistics'
  }
  if (text.includes('義工') || text.includes('物資') || text.includes('社區')) {
    return 'community-support'
  }
  if (text.includes('調查') || text.includes('刑事') || text.includes('貪污')) {
    return 'investigation'
  }
  if (text.includes('政府') || text.includes('官方')) {
    return 'government-announcement'
  }
  return 'general-news'
}

// 添加新聞到 Firestore（使用 news 集合）
async function addNews(news: { title: string; url: string; date: string; description?: string; content?: string; pubDate?: string }) {
  try {
    // 檢查是否已存在
    const exists = await newsExists(news.title, news.url)
    if (exists) {
      console.log(`⏭️  跳過已存在的新聞: ${news.title}`)
      return false
    }
    
    // 使用 description 作為內容，如果沒有則獲取完整內容
    let content = news.description || news.content
    if (!content || content.length < 200) {
      console.log(`📄 正在獲取新聞內容: ${news.title}`)
      try {
        const fullContent = await fetchNewsContent(news.url)
        if (fullContent && fullContent !== '無法獲取新聞內容') {
          content = fullContent
        } else if (!content) {
          content = news.description || '無詳細內容'
        }
      } catch (error: any) {
        console.warn(`獲取完整內容失敗: ${error.message}，使用描述內容`)
        if (!content) {
          content = news.description || '無詳細內容'
        }
      }
    }
    
    // 使用備用分類（本地腳本無法直接調用 Cloud Functions 的 AI 分類）
    const newsCategory = classifyNewsFallback(news.title, content || '')
    
    // 設置標籤
    const tag: 'gov' | 'news' = 'gov' // 政府新聞
    
    // 解析日期和時間
    let timestamp = Timestamp.now()
    try {
      // 優先使用原始 pubDate（包含完整時間信息）
      if (news.pubDate) {
        const parsedDate = new Date(news.pubDate)
        if (!isNaN(parsedDate.getTime())) {
          timestamp = Timestamp.fromDate(parsedDate)
          console.log(`使用 RSS pubDate 解析時間: ${parsedDate.toLocaleString('zh-HK')}`)
        }
      } else {
        // 如果沒有 pubDate，嘗試從格式化的日期字符串解析
        const dateMatch = news.date.match(/(\d{4})年(\d{1,2})月(\d{1,2})日/)
        if (dateMatch) {
          const [, year, month, day] = dateMatch
          const date = new Date(parseInt(year), parseInt(month) - 1, parseInt(day))
          timestamp = Timestamp.fromDate(date)
        }
      }
    } catch (error: any) {
      console.warn(`解析日期時發生錯誤: ${error.message}，使用當前時間`)
    }
    
    const newsItem = {
      title: news.title,
      content: content,
      source: '香港政府新聞公報',
      url: news.url,
      tag,
      newsCategory,
      timestamp
    }
    
    await addDoc(collection(db, 'news'), newsItem)
    console.log(`✅ 已添加新聞: ${news.title} (分類: ${newsCategory})`)
    return true
  } catch (error: any) {
    console.error(`❌ 添加新聞時發生錯誤 (${news.title}):`, error.message)
    return false
  }
}

// 主函數
async function fetchAndAddNews() {
  try {
    // 認證
    await authenticate()
    
    // 獲取新聞
    const newsList = await fetchGovNews()
    
    if (newsList.length === 0) {
      console.log('ℹ️  沒有找到相關的新聞')
      return
    }
    
    console.log('📝 開始處理新聞...\n')
    
    let addedCount = 0
    for (const news of newsList) {
      const added = await addNews(news)
      if (added) {
        addedCount++
      }
      // 添加延遲避免請求過快
      await new Promise(resolve => setTimeout(resolve, 1000))
    }
    
    console.log(`\n✅ 完成！共添加 ${addedCount} 條新公告`)
  } catch (error: any) {
    console.error('\n❌ 執行失敗:', error.message)
    throw error
  }
}

// 執行
fetchAndAddNews()
  .then(() => {
    console.log('\n執行完成')
    process.exit(0)
  })
  .catch((error) => {
    console.error('\n執行失敗:', error)
    process.exit(1)
  })

File: scripts/fetch_gov_news.py
#!/usr/bin/env python3
"""
政府新聞公報獲取器 (Python 版本)
"""

import os
import sys
import re
import time
from datetime import datetime
from typing import List, Dict, Optional
import feedparser
import requests
from bs4 import BeautifulSoup
import firebase_admin
from firebase_admin import credentials, firestore
from dotenv import load_dotenv

# 載入環境變量
load_dotenv()

# 初始化 Firebase
if not firebase_admin._apps:
    # 嘗試使用環境變量中的服務帳戶
    cred_path = os.getenv('GOOGLE_APPLICATION_CREDENTIALS')
    if cred_path and os.path.exists(cred_path):
        cred = credentials.Certificate(cred_path)
        firebase_admin.initialize_app(cred)
    else:
        # 嘗試使用項目根目錄的服務帳戶文件
        default_cred_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'service-account-key.json')
        if os.path.exists(default_cred_path):
            cred = credentials.Certificate(default_cred_path)
            firebase_admin.initialize_app(cred)
        else:
            try:
                # 使用默認憑證（適用於 Cloud Functions 或已設置的環境）
                firebase_admin.initialize_app()
            except Exception as e:
                print("❌ Firebase 初始化失敗！")
                print("\n請設置 Firebase 憑證，方法如下：")
                print("1. 下載服務帳戶密鑰文件（JSON）")
                print("2. 設置環境變量：")
                print("   export GOOGLE_APPLICATION_CREDENTIALS=/path/to/service-account-key.json")
                print("   或將文件放在項目根目錄並命名為 'service-account-key.json'")
                print("\n詳細說明：https://cloud.google.com/docs/authentication/external/set-up-adc")
                raise

db = firestore.client()

# 火災相關關鍵詞（核心關鍵詞，必須包含）
CORE_FIRE_KEYWORDS = [
    "火",
    "火警",
    "火災",
    "火災事故",
    "火災現場",
    "宏福苑",  # 特定地點
]

# 輔助關鍵詞
SUPPORTING_KEYWORDS = [
    "大埔",
    "宏福",
    "庇護中心",
    "臨時庇護",
    "疏散",
    "消防",
    "救援",
    "緊急",
    "撤離",
]


def is_fire_related(text: str) -> bool:
    """檢查文本是否與火災相關"""
    if not text or not text.strip():
        return False
    
    lower_text = text.lower()
    
    # 必須包含至少一個核心關鍵詞
    has_core_keyword = any(
        keyword.lower() in lower_text for keyword in CORE_FIRE_KEYWORDS
    )
    
    if has_core_keyword:
        return True
    
    # 如果沒有核心關鍵詞，檢查是否同時包含多個輔助關鍵詞
    supporting_count = sum(
        1 for keyword in SUPPORTING_KEYWORDS
        if keyword.lower() in lower_text
    )
    
    # 如果包含 2 個或以上的輔助關鍵詞，且包含"大埔"或"宏福"，則認為相關
    if supporting_count >= 2:
        return "大埔" in lower_text or "宏福" in lower_text
    
    return False


def parse_rss_date(pub_date) -> str:
    """解析 RSS pubDate 為中文日期格式"""
    try:
        # feedparser 會自動解析日期為 time.struct_time
        if hasattr(pub_date, 'tm_year'):
            return f"{pub_date.tm_year}年{pub_date.tm_mon}月{pub_date.tm_mday}日"
        elif isinstance(pub_date, str):
            # 如果是字符串，嘗試解析
            dt = datetime.strptime(pub_date[:19], "%Y-%m-%dT%H:%M:%S")
            return f"{dt.year}年{dt.month}月{dt.day}日"
    except:
        pass
    return datetime.now().strftime("%Y年%m月%d日")


def clean_html(html: str) -> str:
    """清理 HTML 標籤和實體"""
    if not html:
        return ""
    
    soup = BeautifulSoup(html, 'html.parser')
    text = soup.get_text()
    
    # 清理多餘空格
    text = re.sub(r'\s+', ' ', text).strip()
    
    return text


def fetch_gov_news() -> List[Dict[str, str]]:
    """獲取政府新聞公報（使用 RSS Feed）"""
    rss_url = "https://www.info.gov.hk/gia/rss/general_zh.xml"
    
    try:
        print(f"📰 正在從 RSS Feed 獲取政府新聞: {rss_url}")
        
        feed = feedparser.parse(rss_url)
        
        if feed.bozo:
            print(f"⚠️  RSS 解析警告: {feed.bozo_exception}")
        
        news_items = []
        
        for entry in feed.entries:
            title = entry.get('title', '').strip()
            link = entry.get('link', '').strip()
            description = entry.get('description', '').strip()
            pub_date = entry.get('published', '')
            
            if not title or not link:
                continue
            
            # 檢查是否與火災相關
            title_related = is_fire_related(title)
            desc_related = is_fire_related(description)
            
            if title_related or desc_related:
                # 使用 published_parsed（feedparser 解析後的日期）或 pub_date
                date_obj = entry.get('published_parsed') or pub_date
                date_str = parse_rss_date(date_obj) if date_obj else datetime.now().strftime("%Y年%m月%d日")
                news_items.append({
                    'title': title,
                    'url': link,
                    'date': date_str,
                    'description': clean_html(description)
                })
                print(f"✅ 找到相關新聞: {title}")
            else:
                print(f"⏭️  跳過不相關新聞: {title}")
        
        print(f"✅ 從 RSS Feed 找到 {len(news_items)} 條相關新聞\n")
        return news_items
        
    except Exception as e:
        print(f"❌ 獲取 RSS Feed 時發生錯誤: {str(e)}")
        raise


def fetch_news_content(url: str) -> str:
    """獲取新聞詳細內容"""
    try:
        response = requests.get(url, headers={
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        }, timeout=10)
        response.raise_for_status()
        
        soup = BeautifulSoup(response.text, 'html.parser')
        
        # 嘗試多種可能的內容選擇器
        content_selectors = [
            '#pressrelease',
            '.pressrelease',
            '#content',
            '.content',
            'article',
            'main'
        ]
        
        content = ""
        for selector in content_selectors:
            element = soup.select_one(selector)
            if element:
                content = element.get_text().strip()
                break
        
        # 如果找不到特定容器，嘗試獲取所有段落
        if not content:
            paragraphs = soup.find_all('p')
            content = '\n\n'.join([
                p.get_text().strip() for p in paragraphs
                if len(p.get_text().strip()) > 20
            ])
        
        return content.strip() or "無法獲取新聞內容"
        
    except Exception as e:
        print(f"獲取新聞內容時發生錯誤 ({url}): {str(e)}")
        return "無法獲取新聞內容"


def announcement_exists(title: str) -> bool:
    """檢查公告是否已存在"""
    try:
        announcements_ref = db.collection('announcements')
        query = announcements_ref.where('title', '==', title).limit(1)
        docs = query.stream()
        return len(list(docs)) > 0
    except Exception as e:
        print(f"檢查公告是否存在時發生錯誤: {str(e)}")
        return False


def add_announcement(news: Dict[str, str]) -> bool:
    """添加公告到 Firestore"""
    try:
        # 檢查是否已存在
        if announcement_exists(news['title']):
            print(f"跳過已存在的公告: {news['title']}")
            return False
        
        # 使用 description 作為內容，如果沒有則獲取完整內容
        content = news.get('description') or news.get('content', '')
        if not content:
            print(f"正在獲取新聞內容: {news['title']}")
            content = fetch_news_content(news['url'])
        
        # 判斷是否為緊急
        is_urgent = is_fire_related(news['title']) and (
            '緊急' in news['title'] or
            '火警' in news['title'] or
            '火災' in news['title'] or
            '緊急' in content or
            '撤離' in content
        )
        
        # 設置標籤
        tag = 'urgent' if is_urgent else 'gov'
        
        # 解析日期
        try:
            date_match = re.match(r'(\d{4})年(\d{1,2})月(\d{1,2})日', news['date'])
            if date_match:
                year, month, day = map(int, date_match.groups())
                timestamp = datetime(year, month, day)
            else:
                timestamp = firestore.SERVER_TIMESTAMP
        except:
            timestamp = firestore.SERVER_TIMESTAMP
        
        announcement = {
            'title': news['title'],
            'content': content,
            'source': '香港政府新聞公報',
            'url': news['url'],
            'isUrgent': is_urgent,
            'tag': tag,
            'timestamp': timestamp
        }
        
        db.collection('announcements').add(announcement)
        print(f"✅ 已添加公告: {news['title']}")
        return True
        
    except Exception as e:
        print(f"添加公告時發生錯誤 ({news['title']}): {str(e)}")
        return False


def fetch_and_add_gov_news():
    """主函數：獲取並添加新聞"""
    try:
        print("📰 開始獲取政府新聞公報...")
        
        # 獲取新聞
        news_list = fetch_gov_news()
        
        if not news_list:
            print("ℹ️  沒有找到相關的新聞")
            return {
                'success': True,
                'added': 0,
                'total': 0,
                'message': '沒有找到相關的新聞'
            }
        
        print(f"📝 開始處理 {len(news_list)} 條新聞...\n")
        
        added_count = 0
        for news in news_list:
            if add_announcement(news):
                added_count += 1
            # 添加延遲避免請求過快
            time.sleep(1)
        
        message = f"處理完成: 新增 {added_count} 條公告，共處理 {len(news_list)} 條新聞"
        print(f"✅ {message}")
        
        return {
            'success': True,
            'added': added_count,
            'total': len(news_list),
            'message': message
        }
        
    except Exception as e:
        print(f"❌ 執行失敗: {str(e)}")
        raise


if __name__ == '__main__':
    try:
        result = fetch_and_add_gov_news()
        print(f"\n執行完成: {result['message']}")
        sys.exit(0)
    except Exception as e:
        print(f"\n執行失敗: {str(e)}")
        sys.exit(1)

